cirq/experiments/random_quantum_circuit_generation.py

```diff
@@ -0,0 +1,226 @@
+    `single_qubit_gates`, but with the constraint that no qubit is acted upon
+    by the same single-qubit gate in consecutive layers. In the layer of
+    two-qubit gates, which pairs of qubits undergo interaction is determined
+    by `pattern`, which is a sequence of two-qubit interaction sets. The
+    set of interactions in a two-qubit layer rotates through this sequence.
+    The two-qubit operations themselves are generated by the call
+    `two_qubit_op_factory(a, b, prng)`, where `a` and `b` are the qubits and
+    `prng` is the pseudorandom number generator.
+
+    At the end of the circuit, an additional layer of single-qubit gates is
+    appended, subject to the same constraint regarding consecutive layers
+    described above.
+    """
+    prng = value.parse_random_state(seed)
+    qubits = list(qubits)
+    coupled_qubit_pairs = _coupled_qubit_pairs(qubits)
+
+    circuit = circuits.Circuit()
+    previous_single_qubit_layer = {}  # type: Dict[cirq.GridQubit, cirq.Gate]
+    for i in range(depth):
+        single_qubit_layer = _single_qubit_layer(qubits, single_qubit_gates,
+                                                 previous_single_qubit_layer,
+                                                 prng)
+        two_qubit_layer = _two_qubit_layer(coupled_qubit_pairs,
+                                           two_qubit_op_factory,
+                                           pattern[i % len(pattern)], prng)
+        circuit.append([g.on(q) for q, g in single_qubit_layer.items()],
+                       strategy=circuits.InsertStrategy.NEW_THEN_INLINE)
+        circuit.append(two_qubit_layer,
+                       strategy=circuits.InsertStrategy.EARLIEST)
+        previous_single_qubit_layer = single_qubit_layer
+    final_single_qubit_layer = _single_qubit_layer(qubits, single_qubit_gates,
+                                                   previous_single_qubit_layer,
+                                                   prng)
+    circuit.append([g.on(q) for q, g in final_single_qubit_layer.items()],
+                   strategy=circuits.InsertStrategy.NEW_THEN_INLINE)
+
+    return circuit
+
+
+def _coupled_qubit_pairs(qubits: List['cirq.GridQubit'],
+                        ) -> List[Tuple['cirq.GridQubit', 'cirq.GridQubit']]:
+    pairs = []
+    qubit_set = set(qubits)
+    for qubit in qubits:
+
+        def add_pair(neighbor: 'cirq.GridQubit'):
+            if neighbor in qubit_set:
+                pairs.append((qubit, neighbor))
+
+        add_pair(devices.GridQubit(qubit.row, qubit.col + 1))
+        add_pair(devices.GridQubit(qubit.row + 1, qubit.col))
+
+    return pairs
+
+
+def _single_qubit_layer(
+        qubits: List['cirq.GridQubit'],
+        single_qubit_gates: Sequence['cirq.Gate'],
+        previous_single_qubit_layer: Dict['cirq.GridQubit', 'cirq.Gate'],
+        prng: 'np.random.RandomState',
+) -> Dict['cirq.GridQubit', 'cirq.Gate']:
```

Comment:
For defense-in-depth, you could check consistency of qubits and previous layer:

```python
if not previous_single_qubit_layer and set(qubits) != set(previous_single_qubit_layer.keys()):
    raise ValueError(...)
```

Reply:
I think you put the `not` there by mistake. I think this is quite unnecessary; I'd rather omit it.